### PR TITLE
Add workgroup swizzling for better cache reuse

### PR DESCRIPTION
**Pull request by nirvedhmeshram**
This PR adds a pass to swizzle workgroup ids for better cache use. 
Currently this is experimental and is done by passing a global option for the log of the swizzle tile size desired. By default the pass does not do any swizzling and will not change the generated code.

### CODE REVIEW

File: iree/compiler/Codegen/LLVMGPU/LLVMGPUSwizzle.cpp
Comment (benvanik) [inline iree/compiler/Codegen/LLVMGPU/LLVMGPUSwizzle.cpp:53]: this is the kind of stuff we'd want on Vulkan and CPU as well, so even if staging it here under LLVMGPU it'd be good to ensure this code stays generic and portable to Codegen/Common/ so that it can be reused.

Comment (nirvedhmeshram) [reply iree/compiler/Codegen/LLVMGPU/LLVMGPUSwizzle.cpp:53]: Makes sense, I can move the pass to Codegen/Common/. This pass can work on any IR after tile and distribute phase. So there is nothing CUDA specific.

File: iree/compiler/Codegen/LLVMGPU/LLVMGPUSwizzle.cpp
Comment (antiagainst) [inline iree/compiler/Codegen/LLVMGPU/LLVMGPUSwizzle.cpp:29]: You'd probably want to generate affine ops. That would compose and optimize better in general.

File: iree/compiler/Codegen/LLVMGPU/LLVMGPUSwizzle.cpp
Comment (antiagainst) [reply iree/compiler/Codegen/LLVMGPU/LLVMGPUSwizzle.cpp:53]: +1. I'm gonna pick this up some days later once #8425 lands. ;-P

File: iree/compiler/Codegen/LLVMGPU/LLVMGPUSwizzle.cpp
Comment (ThomasRaoux) [reply iree/compiler/Codegen/LLVMGPU/LLVMGPUSwizzle.cpp:29]: affine map handle everything as signed and generates extra ops for division/rem, here we definitely want to keep the number of instructions added to a minimum so I’d prefer leave it like that. (Unless there is a way to generate the same code sequence). 
I don’t think there are really opportunities for combining this arithmetic

File: iree/compiler/Codegen/Common/WorkGroupSwizzle.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/Common/WorkGroupSwizzle.cpp:23]: Not required for this PR, but you could make this swizzling a callback and make the core transformation a utility function that takes the call back to get the swizzling. You can use this method as the callback for your use case as a pass in GPU.

File: iree/compiler/Codegen/Common/WorkGroupSwizzle.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/Common/WorkGroupSwizzle.cpp:82]: What is the expectation when `index == 2`?

File: iree/compiler/Codegen/Common/WorkGroupSwizzle.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/Common/WorkGroupSwizzle.cpp:84]: Follow up then does this need to also return if `zFound == true`?

File: iree/compiler/Codegen/Passes.td
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/Passes.td:176]: Not sure why there are two passes. Maybe make one a utility function and another a pass?

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: iree/compiler/Codegen/Common/WorkGroupSwizzle.cpp
Comment (nirvedhmeshram) [reply iree/compiler/Codegen/Common/WorkGroupSwizzle.cpp:82]: We leave index==2 alone if it exists and its not a problem if it doesnt.

File: iree/compiler/Codegen/Common/WorkGroupSwizzle.cpp
Comment (nirvedhmeshram) [reply iree/compiler/Codegen/Common/WorkGroupSwizzle.cpp:84]: No it doesnt make a difference.

File: iree/compiler/Codegen/Passes.td
Comment (nirvedhmeshram) [reply iree/compiler/Codegen/Passes.td:176]: This is only one pass right?

File: iree/compiler/Codegen/Common/WorkGroupSwizzle.cpp
Comment (nirvedhmeshram) [reply iree/compiler/Codegen/Common/WorkGroupSwizzle.cpp:82]: One thing we were careful with this swizzling logic is to not change the shape of the grid. That is why we do not have to worry about what's happening in the third dimension

File: iree/compiler/Codegen/LLVMGPU/Passes.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/LLVMGPU/Passes.cpp:33]: Please rename to `iree-codegen-log-swizzle-tile`.

File: iree/compiler/Codegen/Common/WorkGroupSwizzle.cpp
Comment (MaheshRavishankar) [reply iree/compiler/Codegen/Common/WorkGroupSwizzle.cpp:23]: Could yo add a TODO here that captures that this function could be made a call back and that the core logic in this pass can be made into a utility function?

File: iree/compiler/Codegen/Passes.td
Comment (MaheshRavishankar) [reply iree/compiler/Codegen/Passes.td:176]: Ah sorry my bad....

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Just a few minor comments.

File: iree/compiler/Codegen/Common/WorkGroupSwizzle.cpp
Comment (nirvedhmeshram) [reply iree/compiler/Codegen/Common/WorkGroupSwizzle.cpp:25]: Done.

File: iree/compiler/Codegen/LLVMGPU/Passes.cpp
Comment (nirvedhmeshram) [reply iree/compiler/Codegen/LLVMGPU/Passes.cpp]: Done.

File: iree/compiler/Codegen/Common/WorkGroupSwizzle.cpp
Comment (MaheshRavishankar) [reply iree/compiler/Codegen/Common/WorkGroupSwizzle.cpp:25]: Actually the TODO should be `TODO : Make this a callback and the core functionality in the pass a utility function`.

**Review by MaheshRavishankar — APPROVED** [review]
One minor nit. Please address before submit.

**Comment by nirvedhmeshram** [comment]
@MaheshRavishankar  Fixed, if it looks good, could you please submit? I do not have write access.

**Pull request merged**
